docs/cardano-monads.rst

```diff
@@ -0,0 +1,922 @@
+with the base monads (``Prod`` and ``TestPure``), similarly to what we have now.
+
+The implementation-specific context is extracted in these method definitions.
+
+In case of ``MonadDBGet``, the choice of implementation is determined by the
+base monad (in prod we use rocksdb, in tests we use pure db). However, we also
+have cases where we need runtime configurability, when with the same base monad
+there can be different implementations. For instance, for slotting we can have
+ntp-based slotting or simple slotting, depending on whether the ``--no-ntp`` flag
+was passed. In these cases we choose to use effect sums::
+
+    instance MonadSlotting Prod where
+        type SlottingContext Prod = Either SimpleSlottingVar NtpSlottingVar
+        getCurrentSlot = do
+          sctx <- asks slottingContext
+          either getCurrentSlotSimple getCurrentSlotNtp sctx
+
+While effect context sums are closed, we are fine with it because we don't
+need extensibility at instance definition site, while the implementations
+themselves are defined independently.
+
+Another example is passing ``SendActions``. With the new approach instead of
+``SendActions`` we should be passing ``SendContext`` (that contains the IP of
+the peer, etc), and define ``MonadSend`` to implement the actual sending logic.
+
+Context extensibility
+_____________________
+
+In order to address the issue of extensibility, for each ``Has``-class we can
+define two instances in a generic way::
+
+    class HasFooContext b ctx where
+        fooContext :: ctx -> FooContext b
+
+    instance HasFooContext b (FooContext b, _) where
+        fooContext = fst
+
+    instance {-# OVERLAPPABLE #-} HasFooContext b ctx => HasFooContext b (_, ctx) where
+        fooContext = fooContext . snd
+
+This way it's possible to add a context using ``withReaderT (fooCtx,)``. A
+stylistic issue with this is the use of ``{-# OVERLAPPABLE #-}``, but it seems
+rather benign here. A technical issue is that need ``FooContext`` to be a data
+family (rather than a type family) in order for this to work, but it's a simple
+change.
+
+For instance, consider ``NtpMonad``. Right now we pass ``NtpSlottingVar`` manually
+there. Instead we could add a ``HasNtpSlottingVar ctx`` constraint, and when running
+the methods write ``withReaderT (ntpSlottingVar,)``.
+
+Usage examples from codebase
+____________________________
+
+Finally, now that we know how to define effects with these approach, let's see
+how to use them. Let's take a look at a concrete example, ``usNormalize``::
+
+    usNormalize :: (USLocalLogicMode ctx m) => m ()
+    usNormalize = do
+        tip <- DB.getTip
+        stateVar <- mvState <$> views (lensOf @UpdateContext) ucMemState
+        atomically . writeTVar stateVar =<< usNormalizeDo (Just tip) Nothing
+
+The type signature will change a little, so that we have an explicit ``ReaderT``,
+but an abstract base monad::
+
+    usNormalize :: (USLocalLogicMode ctx b) => ReaderT ctx b ()
```

Comment:
why is `b` needed in `USLocalLogicMode`? it can be deduced from `b`

Reply:
what?

Reply:
if you wanted to say that `ctx` can be deduced from `b` — it can't, that's a crucial part of the approach (the context is decoupled from the base monad)

Reply:
oh, sorry. it seems me confused _global context_ `ctx` and specific to monad `type FooContext b`